File: xllm/core/layers/mlu/rotary_embedding.cpp
/* Copyright 2025 The xLLM Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    https://github.com/jd-opensource/xllm/blob/main/LICENSE

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "rotary_embedding.h"

#include "kernels/ops_api.h"

namespace xllm {
namespace layer {

RotaryEmbeddingImpl::RotaryEmbeddingImpl(int rotary_dim,
                                         int max_position_embeddings,
                                         int rope_theta,
                                         bool interleaved,
                                         const torch::TensorOptions& options)
    : rotary_dim_(rotary_dim),
      max_position_embeddings_(max_position_embeddings),
      rope_theta_(rope_theta),
      interleaved_(interleaved) {
  auto dev_options =
      torch::TensorOptions().device(torch::DeviceType::PrivateUse1);

  auto inv_freq_t = torch::arange(/*start=*/0,
                                  /*end=*/rotary_dim_,
                                  /*step=*/2,
                                  torch::TensorOptions().dtype(torch::kFloat));
  inv_freq_t = inv_freq_t.to(dev_options);
  auto inv_freq =
      1.0 /
      torch::pow(rope_theta_, inv_freq_t / static_cast<double>(rotary_dim_));

  auto t = torch::arange(0, max_position_embeddings_, 1, torch::kFloat32);
  t = t.to(dev_options);

  const auto freqs = torch::einsum("i,j->ij", {t, inv_freq});
  // Create cos and sin embeddings.
  torch::Tensor emd;
  if (interleaved) {
    // [a, b, c, d] => [a, a, b, b, c, c, d, d]
    emd = freqs.repeat_interleave(/*repeats=*/2, /*dim=*/-1);
  } else {
    // [a, b, c, d] => [a, b, c, d, a, b, c, d]
    emd = torch::cat({freqs, freqs}, /*dim=*/-1);
  }

  const auto cos_sin = torch::cat({emd.cos(), emd.sin()}, /*dim=*/-1);
  cos_sin_cache_ = register_buffer("cos_sin_cache", cos_sin.to(options));

  auto cos_sin_vec = cos_sin_cache_.chunk(2, /*dim=*/-1);
  cos_ = cos_sin_vec[0].view({-1, rotary_dim});
  sin_ = cos_sin_vec[1].view({-1, rotary_dim});
}

void RotaryEmbeddingImpl::forward(torch::Tensor& q,
                                  torch::Tensor& k,
                                  const torch::Tensor& positions,
                                  const torch::Tensor& cu_query_lens,
                                  int max_query_len,
                                  bool is_prompt) {
  bool discrete;
  std::optional<torch::Tensor> position_ids;
  if (is_prompt) {
    discrete = false;
    position_ids = std::nullopt;
  } else {
    discrete = true;
    position_ids = positions;
  }

  xllm::kernel::RotaryParams rotary_params;
  rotary_params.q = q;
  rotary_params.k = k;
  rotary_params.sin = sin_;
  rotary_params.cos = cos_;
  rotary_params.cos_sin = cos_sin_cache_;
  rotary_params.position_ids = position_ids;
  rotary_params.cu_query_lens = cu_query_lens;
  rotary_params.interleaved = interleaved_;
  rotary_params.discrete = discrete;
  rotary_params.max_query_len = max_query_len;

  xllm::kernel::apply_rotary(rotary_params);
}

}  // namespace layer
}  // namespace xllm

File: xllm/python/layers/rotary_embedding.py
# Copyright 2025-2026 The xLLM Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://github.com/jd-opensource/xllm/blob/main/LICENSE
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""NEOX-style rotary embedding cache in the layout the fused
``xllm_ops.fused_qk_norm_rope`` kernel expects. Pure table construction — no op
dispatch dependency."""

from __future__ import annotations

import torch
import torch.nn as nn


class RotaryEmbedding(nn.Module):
    """Holds the NEOX-style RoPE cos/sin cache in the exact layout the fused
    ``xllm_ops.fused_qk_norm_rope`` kernel expects.

    ``cos_sin_cache`` is a single ``[max_position, head_dim]`` tensor whose first
    ``head_dim/2`` columns are ``cos(freqs)`` and last ``head_dim/2`` are
    ``sin(freqs)`` (``freqs = outer(positions, inv_freq)``). This matches C++
    ``MRotaryEmbedding::precomputed_cos_sin_cache()`` so both paths use identical
    rotary tables. Built on ``device`` in the model dtype to match C++ exactly.
    """

    def __init__(
        self,
        head_dim: int,
        max_position: int,
        rope_theta: float,
        dtype: torch.dtype | None = None,
        device: torch.device | str | None = None,
    ) -> None:
        super().__init__()
        self.head_dim = head_dim
        inv_freq = 1.0 / (
            rope_theta
            ** (
                torch.arange(0, head_dim, 2, dtype=torch.float32, device=device)
                / head_dim
            )
        )
        t = torch.arange(max_position, dtype=torch.float32, device=device)
        freqs = torch.outer(t, inv_freq)  # [max_position, head_dim/2]
        cos_half = freqs.cos()  # [max_position, head_dim/2]
        sin_half = freqs.sin()
        if dtype is not None:
            cos_half = cos_half.to(dtype)
            sin_half = sin_half.to(dtype)

        device_type = torch.device(device).type if device else "cpu"
        cos_sin_cache = torch.cat([cos_half, sin_half], dim=-1)
        self.register_buffer(
            "cos_sin_cache", cos_sin_cache.contiguous(), persistent=False
        )

    def forward(
        self, positions: torch.Tensor
    ) -> tuple[torch.Tensor, torch.Tensor]:
        """Index into the cache and return per-token (cos, sin).

        Returns cos/sin as ``[1, num_tokens, 1, head_dim]`` via index + view
        (zero allocation on the hot path).
        """
        num_tokens = positions.size(0)
        cos_half = self.cos_sin_cache[positions]
        cos = cos_half.view(1, num_tokens, 1, self.head_dim)
        sin = cos_half.view(1, num_tokens, 1, self.head_dim)
        return cos, sin
